### PR TITLE
Init pre-upgrade and upgrade task page

### PR DESCRIPTION
**Pull request by Monnte**
**OAMG-9574 | OAMG-9572**

Init task page component with additional changes to other part of codebase to stay more consistent and some fixes.

The code is not structured and not typed as should be. Tests not included. 
This stuff will be handled in additional tasks, this commit is initial step for this component to have some base from where we can build :+1:

### CODE REVIEW

File: src/Helpers/constants.ts [hunk 1]
```diff
@@ -1,7 +1,14 @@
 export const PERMISSIONS = {
   readHosts: 'inventory:hosts:read',
+  useTasks: 'tasks:*:*',
 };
 
 export const INVENTORY_API_ROOT = '/api/inventory/v1';
 export const INVENTORY_HOSTS_ROOT = '/hosts';
 export const INVENTORY_TAGS_ROOT = '/tags';
+
+export const TASKS_API_ROOT = '/api/tasks/v1';
```
Comment (andywaltlova) [inline src/Helpers/constants.ts:10]: Same as for api functions, I would continue ith using the prefix, all previous constants start with `INVENTORY_` lets make these `TASKS_`

File: src/Components/TasksPage/RunTaskModal.tsx [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  };
+
+  return (
+    <Modal
+      title={title}
+      isOpen={isOpen}
+      onClose={() => setIsOpen(false)}
+      width={'70%'}
+      actions={[
+        <Button
```
Comment (andywaltlova) [inline src/Components/TasksPage/RunTaskModal.tsx:36]: After this is clicked it would be nice to refresh the table :) (can be separate task in separate PR)

**Review by andywaltlova — CHANGES_REQUESTED** [review]
Looks good to me, after I see the follow up tasks I am fine with merging this base :+1:

File: src/api.ts [hunk 3]
```diff
@@ -17,7 +25,39 @@ const getInventory = async (path: string) => {
   return getResponseOrError(response);
 };
 
-const getResponseOrError = (response) => {
+const getTasks = async (path: string) => {
```
[the remark this answers is not in the document]
Comment (Wecros) [reply src/api.ts:28]: like `tasks_get` here or just `tasksGet`? could be good for code readability. Was also thinking of some type of namespace Inventory / Tasks. But that probably will not fit well with out file structure.

**Review by Wecros — APPROVED** [review]
Overall, I like this table better than the current one in tasks. I feel like the dropdown with extra details is useful, especially the "Systems with alerts" part.

I found a few minor things we could look into in consequent FE issues.

1. The sorting takes more time than in tasks table.
- Not sure why, but in the tasks table the sorting is instantenous for me.
2. I do not like that when clicking on task run details there is 1 second window where it fetches the data.
- I would prefer if the table had prefetched the data when displaying the original table so that the user has instantaneous feedback when clicking on the details of individual task runs.

File: src/api.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (andywaltlova) [reply src/api.ts:28]: just `tasksWhatever` and `inventoryWhatever`, I think it could be could especially in api and helpers where the functions from different part of apps mix together (other options would be to have special api and helpers for each part)

**Comment by andywaltlova** [comment]
> * I would prefer if the table had prefetched the data when displaying the original table so that the user has instantaneous feedback when clicking on the details of individual task runs.

:shrug: I am fan of fetching only what you need (and I am guessing the initial load would be longer), but this is something we can add to our list and ask Andy from UX

File: package.json [hunk 4]
```diff
@@ -27,6 +27,7 @@
     "@redhat-cloud-services/frontend-components-notifications": "^3.2.16",
     "@redhat-cloud-services/frontend-components-utilities": "^3.7.6",
     "classnames": "^2.3.1",
+    "moment": "^2.29.4",
```
[the remark this answers is not in the document]
Comment (Monnte) [reply package.json:30]: Yep, agree, I found out at the same time you wrote that the moment is deprecated :D

File: src/Components/TasksPage/TasksPage.tsx [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            <TasksTable slug={slug} />
+          </GridItem>
+        </Grid>
+        {modalIsOpen && (
+          <RunTaskModal isOpen={modalIsOpen} setIsOpen={setModalIsOpen} title={taskInfo.title} description={taskInfo.description} slug={slug} />
+        )}
+      </React.Fragment>
+    ) : (
+      <TextContent>
+        <Text component="h1">Oops, something went wrong or task is not currently available.</Text>
```
[the remark this answers is not in the document]
Comment (Monnte) [reply src/Components/TasksPage/TasksPage.tsx:62]: :+1:

File: src/Components/TasksPage/TasksPage.tsx [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+export const TasksPage = ({ slug }: TasksPageProps) => {
+  const [[canUseTasks], isLoadingTasks] = useRbac([PERMISSIONS.useTasks], 'tasks');
+
+  const [modalIsOpen, setModalIsOpen] = useState(false);
+  const [taskInfo, setTaskInfo] = useState({} as TaskInfo);
+  const [isTaskAvailable, setTaskAvailable] = useState(false);
+
+  useEffect(() => {
+    fetchTaskInfo(slug).then((fetchedTaskInfo) => {
```
[the remark this answers is not in the document]
Comment (Monnte) [reply src/Components/TasksPage/TasksPage.tsx:25]: Yep, the upgrade task not exists at this time that is why there is oops, and the executed tasks filtering waiting to be fixed :+1:

File: src/Helpers/Helpers.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+export const dispatchNotification = (store, notification) => {
+  store.dispatch(addNotification(notification));
+};
+
+export const isError = (result) => {
+  const statusString = result?.response?.status.toString();
+  return result?.response?.status && statusString[0] !== '2';
+};
+
+export const EXECUTE_TASK_NOTIFICATION = (store, title, ids, task_id) => {
```
[the remark this answers is not in the document]
Comment (Monnte) [reply src/Helpers/Helpers.tsx:14]: Copied from another code base :D I planned the task to fix that kind of stuff, but I think it is easier to fix it right away.

File: src/Components/TasksPage/RunTaskModal.tsx (hunk 2 above)
Comment (Monnte) [reply src/Components/TasksPage/RunTaskModal.tsx:36]: Yep, It is planned :+1:

**Comment by Monnte** [comment]
> Overall, I like this table better than the current one in tasks. I feel like the dropdown with extra details is useful, especially the "Systems with alerts" part.
> 
> I found a few minor things we could look into in consequent FE issues.
> 
> 1. The sorting takes more time than in tasks table.
> 
> * Not sure why, but in the tasks table the sorting is instantenous for me.
> 
> 2. I do not like that when clicking on task run details there is 1 second window where it fetches the data.
> 
> * I would prefer if the table had prefetched the data when displaying the original table so that the user has instantaneous feedback when clicking on the details of individual task runs.

@Wecros 
1. It is because the tasks table fetch all tasks at once and then handles locally sorting etc
2. It would be nice if tasks teams can provide the additional data in the same endpoint as we fetch executed tasks, but for now we need to fetch it one by one :shrug:

**Comment by Monnte** [comment]
All threads should be resolved by the last commit, and also some re-names were made to have consistent files. .tsx upper case, .ts lower case :+1:

File: src/Components/TasksPage/TasksTable.tsx [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+            </Text>
+            <Text component="p" key={`task-details-systems-val-${id}`}>
+              {details.alerts_count}
+            </Text>
+          </TextContent>
+        </GridItem>
+        <GridItem span={12}>
+          <Button
+            variant="secondary"
+            component="a"
```
Comment (andywaltlova) [inline src/Components/TasksPage/TasksTable.tsx:244]:
```suggestion
            component="a"
            icon={<ExternalLinkAltIcon />}
```

image

**Review by andywaltlova — APPROVED** [review]
